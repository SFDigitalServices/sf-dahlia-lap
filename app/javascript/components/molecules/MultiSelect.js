import React from 'react'

import PropTypes from 'prop-types'
import Select from 'react-select'

const getStyle = () => {
  return {
    control: (base, _) => ({
      ...base,
      minHeight: '45px'
    }),
    multiValue: (base, _) => ({
      ...base,
      maxWidth: '90%'
    })
  }
}

const MultiSelect = ({
  input = null,
  options = [],
  value = [],
  disabled = false,
  onChange = () => {}
}) => {
  let selectedValues
  if (value && value[0] && !value[0].value) {
    selectedValues = options.filter((option) => value.includes(option.value))
  } else {
    selectedValues = value
  }
  return (
    <div className='input-custom-height'>
      <Select
        {...input}
        isDisabled={disabled}
        isClearable
        isSearchable
        isMulti
        styles={getStyle()}
        onChange={(values) => onChange(values)}
        options={options}
        value={selectedValues}
      />
    </div>
  )
}

const MultiSelectItemShape = PropTypes.shape({
  value: PropTypes.string,
  label: PropTypes.string
})

MultiSelect.propTypes = {
  input: PropTypes.object,
  options: PropTypes.arrayOf(MultiSelectItemShape),
  disabled: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.arrayOf(MultiSelectItemShape),
    PropTypes.arrayOf(PropTypes.string)
  ]),
  onChange: PropTypes.func
}

export { MultiSelect as default, MultiSelectItemShape }
