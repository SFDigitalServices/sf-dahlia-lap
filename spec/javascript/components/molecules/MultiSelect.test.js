import React from 'react'

import { shallow } from 'enzyme'
import Select from 'react-select'

import MultiSelect from 'components/molecules/MultiSelect'

const mockItems = [
  {
    value: 'item1',
    label: 'Item1'
  },
  {
    value: 'item2',
    label: 'Item2'
  },
  {
    value: 'item3',
    label: 'Item3'
  }
]

describe('MultiSelect', () => {
  describe('when items is null', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect />)
    })

    test('should render no items or selectedItems', () => {
      expect(wrapper.find(Select).props().value).toEqual([])
      expect(wrapper.find(Select).props().options).toEqual([])
    })
  })

  describe('with multiple items', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect items={mockItems} />)
    })

    test('renders no selected items', () => {
      expect(wrapper.find(Select).props().value).toEqual([])
    })

    test('renders the items', () => {
      expect(wrapper.find(Select).props().options).toEqual(mockItems)
    })

    test('is not disabled', () => {
      expect(wrapper.find(Select).props().isDisabled).toBeFalsy()
    })
  })

  describe('when disabled', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect items={mockItems} disabled />)
    })

    test('renders as disabled', () => {
      expect(wrapper.find(Select).props().isDisabled).toBeTruthy()
    })
  })

  describe('when one item is selected', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect items={mockItems} selectedItems={[mockItems[0]]} disabled />)
    })

    test('should pass that item to react-select', () => {
      expect(wrapper.find(Select).props().value).toEqual([mockItems[0]])
    })
  })

  describe('with default height', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect items={mockItems} />)
    })

    test('sets height to 45px', () => {
      expect(wrapper.find(Select).props().styles.control().height).toEqual('45px')
    })
  })

  describe('with small height', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<MultiSelect items={mockItems} height='small' />)
    })

    test('sets height to 38px', () => {
      expect(wrapper.find(Select).props().styles.control().height).toEqual('38px')
    })
  })
})
