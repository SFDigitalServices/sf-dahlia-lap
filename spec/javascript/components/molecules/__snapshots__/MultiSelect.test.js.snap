// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect when disabled renders as disabled 1`] = `
<DocumentFragment>
  <div
    class="input-custom-height"
  >
    <div
      class=" css-3iigni-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-6-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
      />
      <div
        aria-disabled="true"
        class=" css-14pefp6-control"
      >
        <div
          class=" css-1fdsijx-ValueContainer"
        >
          <div
            class=" css-1jqq78o-placeholder"
            id="react-select-6-placeholder"
          >
            Select...
          </div>
          <div
            class=" css-1h01tm3-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-describedby="react-select-6-placeholder"
              aria-expanded="false"
              aria-haspopup="true"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class=""
              disabled=""
              id="react-select-6-input"
              role="combobox"
              spellcheck="false"
              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            class=" css-894a34-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            class=" css-1xc3v61-indicatorContainer"
          >
            <svg
              aria-hidden="true"
              class="css-tj5bde-Svg"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MultiSelect with default height sets height to 45px 1`] = `
<DocumentFragment>
  <div
    class="input-custom-height"
  >
    <div
      class=" css-b62m3t-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-8-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
      />
      <div
        class=" css-1vyczjh-control"
      >
        <div
          class=" css-1fdsijx-ValueContainer"
        >
          <div
            class=" css-1jqq78o-placeholder"
            id="react-select-8-placeholder"
          >
            Select...
          </div>
          <div
            class=" css-qbdosj-Input"
            data-value=""
          >
            <input
              aria-autocomplete="list"
              aria-describedby="react-select-8-placeholder"
              aria-expanded="false"
              aria-haspopup="true"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              class=""
              id="react-select-8-input"
              role="combobox"
              spellcheck="false"
              style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            class=" css-1u9des2-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            class=" css-1xc3v61-indicatorContainer"
          >
            <svg
              aria-hidden="true"
              class="css-tj5bde-Svg"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
